File: day81-90/day82/951. Flip Equivalent Binary Trees.py
from collections import deque
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BinaryTree:
    
    def build_tree(self, arr):
        if not arr:
            return None
        
        
        root = TreeNode(arr[0])
        queue = deque([root])
        index = 1
        
        
        while queue and index < len(arr):
            node = queue.popleft()
            
            
            if arr[index] is not None:
                node.left = TreeNode(arr[index])
                queue.append(node.left)
            index += 1
            
            
            if index >= len(arr):
                break
            
            
            if arr[index] is not None:
                node.right = TreeNode(arr[index])
                queue.append(node.right)
            index += 1
        
        return root
    
    
    def print_tree(self,root):
        if not root:
            return "Empty Tree"
        
        queue = deque([root])
        result = []
        while queue:
            node = queue.popleft()
            if node:
                result.append(node.val)
                queue.append(node.left)
                queue.append(node.right)
            else:
                result.append(None)
        
        
        while result and result[-1] is None:
            result.pop()
        
        return result
class Solution:
    def flipEquiv(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        if not root1 and not root2:
            return True
        if not root1 or not root2:
            return False
        if root1.val!=root2.val:
            return False
        return (
             self.flipEquiv(root1.left, root2.left) and self.flipEquiv(root1.right, root2.right) 
             or 
             self.flipEquiv(root1.left, root2.right) and self.flipEquiv(root1.right, root2.left))
        
bt=BinaryTree()        
#root1=bt.build_tree([1,2,3,4,5,6,None,None,None,7,8])
#root2=bt.build_tree([1,3,2,None,6,4,5,None,None,None,None,8,7])
root1=bt.build_tree([])
root2=bt.build_tree([1])
# print(bt.print_tree(root1))
# print(bt.print_tree(root2))

sol=Solution()
print(sol.flipEquiv(root1,root2))

File: day81-90/day82/951. Flip Equivalent Binary Trees.php
<?php

class TreeNode {
    public $val;
    public $left;
    public $right;
    
    public function __construct($val = 0, $left = null, $right = null) {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

class BinaryTree {
    public $root;

    

    public function buildTree($arr) {
        if (empty($arr)) {
            return null;
        }

        $root = new TreeNode($arr[0]);
        $queue = new SplQueue();
        $queue->enqueue($root);
        $index = 1;

        while (!$queue->isEmpty() && $index < count($arr)) {
            $node = $queue->dequeue();

            
            if ($arr[$index] !== null) {
                $node->left = new TreeNode($arr[$index]);
                $queue->enqueue($node->left);
            }
            $index++;

            if ($index >= count($arr)) {
                break;
            }

            
            if ($arr[$index] !== null) {
                $node->right = new TreeNode($arr[$index]);
                $queue->enqueue($node->right);
            }
            $index++;
        }

        return $root;
    }

    public function printTree($root) {
        if (!$root) {
            return "Empty Tree";
        }

        $queue = new SplQueue();
        $queue->enqueue($root);
        $result = [];

        while (!$queue->isEmpty()) {
            $node = $queue->dequeue();
            if ($node) {
                $result[] = $node->val;
                $queue->enqueue($node->left);
                $queue->enqueue($node->right);
            } else {
                $result[] = null;
            }
        }

        
        while (!empty($result) && end($result) === null) {
            array_pop($result);
        }

        return $result;
    }
}



class Solution{
    /**
     * @param TreeNode $root1
     * @param TreeNode $root2
     * @return Boolean
     */
    function flipEquiv($root1, $root2) {
        if (!$root1 && !$root2) {
            return true;
        }
        if (!$root1 || !$root2) {
            return false;
        }
        if ($root1->val != $root2->val) {
            return false;
        }
        return (
            $this->flipEquiv($root1->left, $root2->right) && $this->flipEquiv($root1->right, $root2->left) 
            || 
            $this->flipEquiv($root1->left, $root2->left) && $this->flipEquiv($root1->right, $root2->right));
    }
}

$binaryTree=new BinaryTree();
$root1=$binaryTree->buildTree([1,2,3,4,5,6,null,null,null,7,8]);
$root2=$binaryTree->buildTree([1,3,2,null,6,4,5,null,null,null,null,8,7]);
$solution=new Solution();
print($solution->flipEquiv($root1,$root2));